tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/WeightedSnapshotWithExemplars.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.tritium.metrics.registry;
+
+import com.codahale.metrics.Snapshot;
+import com.codahale.metrics.WeightedSnapshot;
+import com.codahale.metrics.WeightedSnapshot.WeightedSample;
+import com.google.common.collect.ImmutableList;
+import java.io.OutputStream;
+import java.util.Collection;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link WeightedSnapshot} with support for storing exemplar metadata for each sample.
+ */
+final class WeightedSnapshotWithExemplars extends Snapshot implements ExemplarsCapture {
+
+    /**
+     * A single sample item with value, weights and optional exemplar metadata.
+     */
+    static class WeightedSampleWithExemplar {
+
+        private final long value;
+        private final double weight;
+        private final @Nullable Object exemplarMetadata;
+
+        WeightedSampleWithExemplar(long value, double weight, @Nullable Object exemplarMetadata) {
+            this.value = value;
+            this.weight = weight;
+            this.exemplarMetadata = exemplarMetadata;
+        }
+
+        public long getValue() {
+            return value;
+        }
+
+        public double getWeight() {
+            return weight;
+        }
+
+        @Nullable
+        public Object getExemplarMetadata() {
```

Comment:
style nit: no need to mark these public since they're not overriding an interface method, package private is sufficient. I'd prefer to match the record style with getters like `value()` and `weight()` as opposed to `getValue()` and `getWeight()`/etc